Division by zero in OneVsRest's `predict_proba` (see #31224); bugfix for `test_ovo_decision_function`

Luis-Varona opened the pull request:
When a classifier has zero confidence in a class, zero rows appear in the predicted probability matrix, resulting in undefined behavior during probability normalization. OneVsRestClassifier's predict_proba method attempts to divide by zero row sums:

```
        .... (precursor logic)
        if not self.multilabel_:
            # Then, probabilities should be normalized to 1.
            Y /= np.sum(Y, axis=1)[:, np.newaxis]
        return Y
```

This fixes that behavior by setting all zero row sums to 1:
```
        .... (precursor logic)
        if not self.multilabel_:
            # Then, (nonzero) probabilities should be normalized to 1.
            row_sums = np.sum(Y, axis=1)[:, np.newaxis]
            row_sums[row_sums == 0] = 1 # Avoid division by zero
            Y /= row_sums
        return Y
```

See Issue #31224 for more details.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Luis-Varona committed:
Fix division by zero in OneVsRestClassifier's predict_proba method

When a classifier has zero confidence in a class, zero rows appear in the predicted probability matrix, resulting in undefined behavior during probability normalization. OneVsRestClassifier's predict_proba method attempts to divide by zero row sums. This fixes that behavior by setting all zero row sums to 1.

jeremiedbb commented on sklearn/multiclass.py:
```diff
@@ -553,8 +553,10 @@ def predict_proba(self, X):
-            # Then, probabilities should be normalized to 1.
-            Y /= np.sum(Y, axis=1)[:, np.newaxis]
+            # Then, (nonzero) probabilities should be normalized to 1.
+            row_sums = np.sum(Y, axis=1)[:, np.newaxis]
+            row_sums[row_sums == 0] = 1 # Avoid division by zero
+            Y /= row_sums
```
```suggestion
            row_sums = np.sum(Y, axis=1)[:, np.newaxis]
            np.divide(Y, row_sums, out=Y, where=row_sums != 0)
```

MarcBresson replied:
that's very elegant!

jeremiedbb reviewed (commented):
Thanks for the PR @Luis-Varona. Let's use numpy's divide function which allows to specify where to divide.

Please add a test in `sklearn/tests/test_multiclass.py` and an entry in the changelog (see https://github.com/scikit-learn/scikit-learn/blob/main/doc/whats_new/upcoming_changes/README.md).

Luis-Varona commented:
Will do. 🙂

Luis-Varona committed:
Refactor div-by-zero fix; update changelog, tests

Refactored the division-by-zero fix in OneVsRest's predict_proba to use np.divide. Documented the change in the changelog. Added test_ovr_single_label_predict_proba_zero_row in tests to validate the bug fix.

Luis-Varona commented:
@jeremiedbb @MarcBresson I've made the requested changes. (The only thing I'm worried about is whether `test_ovr_single_label_predict_proba_zero_row` is too long of a test function name; we already have `test_ovr_single_label_predict_proba`, and I wanted to extract the logic into a separate function for obvious reasons.) Let me know if everything is good 🙂

Luis-Varona committed:
Merge branch 'main' into 31224-onevsrest-div-by-zero

Luis-Varona committed:
Reformatted code with ruff

Luis-Varona committed:
Merge branch 'main' into 31224-onevsrest-div-by-zero

MarcBresson commented on sklearn/tests/test_multiclass.py:
```diff
@@ -408,6 +409,36 @@ def test_ovr_multilabel_predict_proba():
         assert_array_equal(pred, Y_pred)
 
 
+def test_ovr_multilabel_predict_proba_zero_row():
```
I'm not sure to understand what this test is actually testing. Could you explain it to me please?

Luis-Varona replied:
That... is probably not supposed to be there. (I later renamed it to `test_ovr_single_label_predict_proba_zero_row`.) My bad, I'll remove it.

MarcBresson commented on sklearn/tests/test_multiclass.py:
```diff
@@ -429,6 +460,39 @@ def test_ovr_single_label_predict_proba():
+
+    base_clf = NaiveBinaryClassifier()
+    X, Y = iris.data, iris.target  # Three-class problem with 150 samples
+    zero_indices = np.random.choice(np.arange(80, len(X)), size=5, replace=False)
+    X[zero_indices] = 0  # Change 5 random samples in the test set to be all zeros
+
+    X_train, Y_train = X[:80], Y[:80]
+    X_test = X[80:]
```
Nitpicking: You could also chose the zero_indices directly from `X_test`

Luis-Varona replied:
I specifically wanted to make only some rows of `X_test` zero, as I wanted both nonzero rows (summing to 1 after processing in `Y_proba`) and zero rows (summing to 0 after processing in `Y_proba`). If I made everything 0, then we wouldn't be able to validate that it works when `X_test` is a mix of both types. Thoughts?

MarcBresson replied:
Sorry for being unclear. I meant that you could have done

```python
    X_train, Y_train = X[:80], Y[:80]
    X_test = X[80:]

    zero_indices = np.random.choice(np.arange(len(X_test)), size=5, replace=False)
    X_test[zero_indices] = 0  # Change 5 random samples in the test set to be all zeros
```

and then remove `zero_indices -= 80`

Luis-Varona commented on sklearn/tests/test_multiclass.py:
```diff
@@ -408,6 +409,36 @@ def test_ovr_multilabel_predict_proba():
+    assert_array_equal(pred, Y_pred)
+
+    zero_idx -= 80
+    assert np.all(Y_pred[zero_idx] == 0)
```
As I said in another comment, this test is not meant to exist. My bad :)

Luis-Varona commented on sklearn/tests/test_multiclass.py:
```diff
@@ -429,6 +460,39 @@ def test_ovr_single_label_predict_proba():
     assert not (pred - Y_pred).any()
 
 
+def test_ovr_single_label_predict_proba_zero_row():
```
Thank you!

Luis-Varona committed:
Removed leftover test; refactored zero indexing

Removed test (test_ovr_multilabel_predict_proba_zero_row) accidentally left over from prototyping. Changed zero indexing logic in test_ovr_single_label_predict_proba_zero_row to be more streamlined.

MarcBresson commented:
There is one test failing in https://github.com/scikit-learn/scikit-learn/actions/runs/14620590689/job/41019410025?pr=31228

` FAILED tests/test_multiclass.py::test_ovo_decision_function - assert 145 > 146`

Luis-Varona commented:
> There is one test failing in https://github.com/scikit-learn/scikit-learn/actions/runs/14620590689/job/41019410025?pr=31228
> 
> ` FAILED tests/test_multiclass.py::test_ovo_decision_function - assert 145 > 146`

@MarcBresson this is not my test; I didn't touch it. Presumably if we want to fix this, it should be in another Issue/PR?

Luis-Varona committed:
Merge branch 'main' into 31224-onevsrest-div-by-zero

Luis-Varona committed:
Lowered threshold in test_ovo_decision_function

Lowered the threshold of unique values in the iris dataset from >146 to >140 in the test_ovo_decision_function test of the test_multiclass suite. Previous CI tests ended up with 145 unique values, resulting in a failing 'assert 145 > 146' check.

Luis-Varona commented:
@MarcBresson @jeremiedbb I fixed the `test_ovo_decision_function` check; it was an overly strict requirement on the `OneVsOneClassifier` decision function. (It was not strictly related in logic to our issues with the `OneVsRestClassifier` class, but @MarcBresson requested the change, so I made it, if that makes sense. 🙂) All CI tests are now passing; it should be ready to merge, I think, unless you have any further feedback?

Luis-Varona committed:
Merge branch 'main' into 31224-onevsrest-div-by-zero

Luis-Varona commented:
@MarcBresson @jeremiedbb, just wanted to follow up and see if there are any more changes you’d like me to make before merging 🙂

MarcBresson commented:
Hello @Luis-Varona , I'm not a maintainer, just a happy contributor :)

Reviews from the scikit-learn team can be quite long, as they have a lot on their plate. But don't worry, it will come!

jeremiedbb committed:
simplify test

jeremiedbb committed:
revert unrelated

Luis-Varona commented:
@jeremiedbb Saw you added some commits; anything else you'd like me to do? 🙂

jeremiedbb reviewed (approved):
I just simplified a bit the test. LGTM. Thanks @Luis-Varona and @MarcBresson !
